src/execution/vm/llvm_engine.cpp

```diff
@@ -0,0 +1,1125 @@
+        TPL_ASSERT(blocks[branch_target_bb_pos] != nullptr, "Branch target does not point to valid basic block");
+
+        auto *check = llvm::ConstantInt::get(type_map()->Int8Type(), 1, false);
+        llvm::Value *cond = ir_builder->CreateICmpEQ(args[0], check);
+
+        if (bytecode == Bytecode::JumpIfTrue) {
+          ir_builder->CreateCondBr(cond, blocks[branch_target_bb_pos], blocks[fallthrough_bb_pos]);
+        } else {
+          ir_builder->CreateCondBr(cond, blocks[fallthrough_bb_pos], blocks[branch_target_bb_pos]);
+        }
+        break;
+      }
+
+      case Bytecode::Return: {
+        if (FunctionHasDirectReturn(func_info.func_type())) {
+          llvm::Value *ret_val = locals_map.GetArgumentById(func_info.GetReturnValueLocal());
+          ir_builder->CreateRet(ir_builder->CreateLoad(ret_val));
+        } else {
+          ir_builder->CreateRetVoid();
+        }
+        break;
+      }
+
+      default: {
+        //
+        // In the default case, each bytecode makes a function call into its
+        // bytecode handler function.
+        //
+
+        llvm::Function *handler = LookupBytecodeHandler(bytecode);
+        issue_call(handler, args);
+        break;
+      }
+    }
+
+    //
+    // If the next bytecode marks the start of a new basic block, we need to
+    // switch insertion points to it before continuing IR generation
+    //
+
+    auto next_bytecode_pos = iter.GetPosition() + iter.CurrentBytecodeSize();
+
+    if (auto blocks_iter = blocks.find(next_bytecode_pos); blocks_iter != blocks.end()) {
+      if (!Bytecodes::IsJump(bytecode) && !Bytecodes::IsTerminal(bytecode)) {
+        ir_builder->CreateBr(blocks_iter->second);
+      }
+      ir_builder->SetInsertPoint(blocks_iter->second);
+    }
+  }
+}
+
+void LLVMEngine::CompiledModuleBuilder::DefineFunctions() {
+  //
+  // We iterate over all the TPL functions defined in the module and generate
+  // their LLVM equivalents into the current module.
+  //
+
+  llvm::IRBuilder<> ir_builder(context());
+  for (const auto &func_info : tpl_module().functions()) {
+    DefineFunction(func_info, &ir_builder);
+  }
+}
+
+void LLVMEngine::CompiledModuleBuilder::Verify() {
+  std::string result;
+  llvm::raw_string_ostream ostream(result);
+  if (bool has_error = llvm::verifyModule(*module(), &ostream); has_error) {
+    // TODO(pmenon): Do something more here ...
+    EXECUTION_LOG_ERROR("ERROR IN MODULE:\n{}", ostream.str());
+  }
+}
+
+void LLVMEngine::CompiledModuleBuilder::Simplify() {
+  //
+  // This function ensures all bytecode handlers marked 'always_inline' are
+  // inlined into the main TPL program. After this inlining, we clean up any
+  // unused functions.
+  //
+
+  llvm::legacy::PassManager pass_manager;
+  pass_manager.add(llvm::createAlwaysInlinerLegacyPass());
+  pass_manager.add(llvm::createGlobalDCEPass());
+  pass_manager.run(*module());
+}
+
+void LLVMEngine::CompiledModuleBuilder::Optimize() {
```

Comment:
I appreciate the small steps approach, but do we at least have the right API in place so that incremental/configurable optimization levels doesn’t require a major refactor in the future?